File: Vindeep-Chaudhari/414_Third_Maximum_Number.cpp
/**
 * Problem Link - https://leetcode.com/problems/third-maximum-number/
 */

class Solution {
public:
    int thirdMax(vector<int>& nums) {
        set<int> s;
        for(int i = 0; i < nums.size(); i++) {
            s.insert(nums[i]);
            if(s.size() > 3) {
                s.erase(s.begin());
            }
        }
        
        if(s.size() == 3) {
            return *s.begin();
        }
        
        return *s.rbegin();
    
    }
};

File: Lokesh_Saana/414-third-maximum-number.py
# https://leetcode.com/problems/third-maximum-number/
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        x = defaultdict(int)
        for i in nums:
            x[i] += 1
        x = sorted(x.items(), key=lambda a: (a[0], a[1]))
        if len(x) < 3:
            return max(x)[0]
        else:
            return x[-3][0]
